astropy/units/quantity_helper/function_helpers.py

```diff
@@ -438,6 +440,67 @@ def _block(arrays, max_depth, result_ndim, depth=0):
         return np_core.shape_base._atleast_nd(arrays, result_ndim)
 
 
+if NUMPY_LT_2_0:
+    @dispatched_function
+    def arange(*args, start=None, stop=None, step=None, dtype=None, like=None):
+        return arange_impl(*args, start=start, stop=stop, step=step, dtype=dtype, like=like)
+else:
+    @dispatched_function
+    def arange(*args, start=None, stop=None, step=None, dtype=None, device=None, like=None):
+        return arange_impl(*args, start=start, stop=stop, step=step, dtype=dtype, device=device, like=like)
+
+def arange_impl(*args, start=None, stop=None, step=None, dtype=None, device=None, like=None):
+    from astropy.units import Quantity
+
+    # NumPy is supposed to validate the input parameters before this dispatched
+    # function is reached. Nevertheless, we'll sprinkle a few rundundant
+    # sanity checks in the form of `assert` statements.
+    # As they are not part of the business logic, it is fine if they are
+    # compiled-away (e.g. the Python interpreter runs with -O)
+    assert like is None
+
+    # now we can bind positional arguments to their meaningful names
+    assert len(args) <= 4
+    match args:
```

Comment:
I think this is the first time the requirements for an implementation are so complex that using `match/case` feels simpler than the equivalent forest of if/else , but I'm *really* glad we have it.